src/shared/udev-util.c

```diff
@@ -319,3 +322,50 @@ bool device_for_action(sd_device *dev, DeviceAction action) {
 
         return a == action;
 }
+
+int udev_rule_parse_value(char *str, char **ret_value, char **ret_endpos) {
+        char *value = str;
+        char *i, *j;
+        int r;
+        bool is_escaped;
+
+        /* value must be double quotated */
+        is_escaped = value[0] == 'e';
+        value += is_escaped;
+        if (value[0] != '"')
+                return -EINVAL;
+        value++;
+
+        if (!is_escaped) {
+                /* unescape double quotation '\"'->'"' */
+                for (i = j = value; *i != '"'; i++, j++) {
+                        if (*i == '\0')
+                                return -EINVAL;
+                        if (i[0] == '\\' && i[1] == '"')
+                                i++;
+                        *j = *i;
+                }
+                j[0] = '\0';
+        } else {
+                _cleanup_free_ char *unescaped = NULL;
+
+                /* find the end position of value */
+                for (i = value; *i != '"'; i++) {
+                        if (i[0] == '\\')
+                                i++;
+                        if (*i == '\0')
+                                return -EINVAL;
+                }
+                i[0] = '\0';
+
+                r = cunescape_length(value, i - value, 0, &unescaped);
+                if (r < 0)
+                        return r;
+                assert(r <= i - value);
+                memcpy(value, unescaped, r + 1);
```

Comment:
I can't say this makes me very happy. It'd be more usual to just pass back the allocated string. But OK, I see that that'd require restructuring the surrounding code.

Reply:
Or maybe a cunescape_inplace? I think the escaped string always takes even or more space than the original (unescaped) one

Reply:
Yes, the unescaped must always be shorter.

I don't think it makes sense to complicate things... We do an extra allocation, but I don't think it matters too much. Parsing is always slow. And the extra memcpy is very fast. Nowadays we generally try to not hyper-optimize everything and instead create new "objects" whenever doing transformations. This makes the code easier to understand and to catch bugs and reuse code. But here the original string was already being modified, so I think the current approach in this patch is OK.

Reply:
I think in-place escaping doesn't look that bad if I just move the body of another function out https://github.com/afq984/systemd/commit/0fbea053fbc8fec9cdbd921206b2e06f62ba1004. But I don't know if it's worth the risk. The only benefit here is probably making the malloc-and-memcpy-back look less weird.